### PR TITLE
HMR

### PR DESCRIPTION
**Pull request by zamotany**
# How to enable
* Add `import 'haul-hmr';` to the file with a root component - usually it's `index.ios.js` or `index.android.js`.
* From within _Developer menu_ press _Enable Hot Reloading_.

# Todo
- [x] Nice API
- [x] Support for multi root component projects
- [x] Manual setup guide
- [x] `react-native-navigation` setup guide
- [x] `react-navigation` setup guide
- [ ] `native-navigation` setup guide
- [x] API docs
- [x] merge latest `master` when it's stable - #199 #196 
- [x] tested & working with `react-native-navigation`
- [x] tested & working with `react-navigation`
- [ ] tested & working with `native-navigation`
# Know issues
- [x] hot middleware sometimes throw an error and crashes haul
- [x] ~paths in `module.hot.accept` are invalid is some cases~
- [x] ~path in `require.resolve` and `require` are not always valid~ (babel passes absolue paths, but they gets shorten by webpack)

### CODE REVIEW

File: src/utils/makeReactNativeConfig.js [hunk 1]
```diff
@@ -197,6 +213,10 @@ const getDefaultConfig = (
     mainFields: ['react-native', 'browser', 'main'],
     extensions: [`.${platform}.js`, '.native.js', '.js'],
   },
+  /**
+   * Set target to webworker as it's closer to RN environment than `web`.
+   */
+  target: 'webworker',
```
Comment (satya164) [inline src/utils/makeReactNativeConfig.js]: What are the list of differences between this target and web?

Comment (zamotany) [inline src/utils/makeReactNativeConfig.js]: `web` assumes you have the DOM, `webworker` doesn't have the DOM, so it uses different mechanism for downloading files - `importScripts` whereas `web` uses `script` tags.

Downloading hot update mechanism for `web`: https://github.com/webpack/webpack/blob/master/lib/JsonpMainTemplate.runtime.js#L12-L18

And the same for `webworker`: https://github.com/webpack/webpack/blob/master/lib/webworker/WebWorkerMainTemplate.runtime.js#L12-L14

React Native is closer to `webworker` in terms of environment than `web`.

Comment (satya164) [inline src/utils/makeReactNativeConfig.js]: > it uses different mechanism for downloading files

nice

> there are other differences in other files

what are the other differences? do they impact anything else than script loading mechanism?

Comment (zamotany) [inline src/utils/makeReactNativeConfig.js]: Different files with templates 😛 I think plugins and loader might behave differently. But it's possible that plugins and loaders get templates from `compiler` / `compilation` so in theory everything should be fine.

Comment (zamotany) [inline src/utils/makeReactNativeConfig.js]: I've check all templates for `web` and `webworker` and they seems to differ only in terms of script/files loading.

**Comment by zamotany** [comment]
@grabbou @satya164 I can also write babel plugin to incject `module.hot` code at the bottom automatically and tweak 2nd argument of `registerComponent`, so the user will have to just `import withHmr from 'haul-hmr';`. WDYT??

File: src/utils/makeReactNativeConfig.js (hunk 1 above)
Comment (satya164) [inline src/utils/makeReactNativeConfig.js]: cool

**Comment by satya164** [comment]
> I can also write babel plugin to incject

sounds good

**Comment by zamotany** [comment]
Added babel plugin so now to enable HMR just move `AppRegistry.registerComponent` to separate file and `import 'haul-hmr'` and that's all!

**Comment by grabbou** [comment]
This pull request is pending on my list of tasks to go over. Will update its status soon (ETA this week).

File: src/HMR/babelPlugin.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
+/**
+ * Copyright 2017-present, Callstack.
+ * All rights reserved.
+ */
+
+/**
+ * Ideally, this file SHOULD NOT be processed by prettier, since trailing commas
+ * are not supported and will cause errors in bundle. This file is used by `babel-loader`
```
Comment (satya164) [inline src/HMR/babelPlugin.js]: > since trailing commas are not supported and will cause errors in bundle

Should we use the 'es5' option for trailing commas

**Comment by satya164** [comment]
https://twitter.com/iamdevloper/status/397664295875805184

File: src/HMR/babelPlugin.js (hunk 2 above)
Comment (zamotany) [inline src/HMR/babelPlugin.js]: That's not for me to decide

Comment (satya164) [inline src/HMR/babelPlugin.js]: @grabbou thou are summoned

Comment (satya164) [inline src/HMR/babelPlugin.js]: My personal opinion is yes, because not all of the code we write will be transpiled. So it'll be nice to have an eslint config which works for every project.

**Comment by satya164** [comment]
Just a small thing, instead of `import 'haul-hmr'`, can we do `import 'haul/hmr'`, where `hmr.js` is a empty file. So that ESLint doesn't show an error for unresolved module.

File: src/HMR/client/utils.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright 2017-present, Callstack.
+ * All rights reserved.
+ */
+
+/* @flow */
+
+// $FlowFixMe
+import Platform from 'Platform'; // eslint-disable-line import/no-extraneous-dependencies
```
Comment (satya164) [inline src/HMR/client/utils.js]: Can we do `import { Platform } from 'react-native'` instead, since we discourage haste, we shouldn't be using it

It'll also remove the eslint and flow comments

File: README.md [hunk 4]
```diff
@@ -67,6 +68,9 @@ Check out the docs to learn more about available commands and tips on customizin
 1. [Configuration](docs/Configuration.md)
 1. [Recipes](docs/Recipes.md)
 
+### Hot Module Replacement
+To enable HMR, add `import 'haul-hmr'` to the file with a root component (usually it's `index.ios.js` or `index.android.js`), then from within the _Developer menu_ press _Enable Hot Reloading_.
```
Comment (satya164) [inline README.md]: Instructions seem to be wrong (?). Let's link to the docs here

File: docs/hmr/Setup.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
+# Hot Module Replacement setup
+Jump to: [Automatic setup](#autoamtic-setup) | [Manual setup](#manual-setup)
```
Comment (satya164) [inline docs/hmr/Setup.md]: Typo: `autoamtic`

File: docs/hmr/Setup.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+For smaller project, we provider automatic option for setting up Hot Module Replacement. If your project satisfy the following criteria, you're good to go to use automatic setup:
```
Comment (satya164) [inline docs/hmr/Setup.md]: I'd rephrase to

> `If your project meets the following criteria, you can use our one-line setup for Hot Module Replacement:`

File: docs/hmr/Setup.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+* Signle root component exported from the root file with `default` keyword (`export default`)
```
Comment (satya164) [inline docs/hmr/Setup.md]: Typo: `Signle`

File: docs/hmr/Setup.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+export default class MyProject extends Component {
+  /* ... */
+}
+
+/* ... */
+
+AppRegistry.registerComponent('MyProject', () => MyProject);
+```
+
+All **default** `index.ios.js` and `index.android.js` files created by `react-native init` are fine for automatic setup.
```
Comment (satya164) [inline docs/hmr/Setup.md]: `are fine for` -> `work with the`

File: docs/hmr/Setup.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+## Automatic setup
+
+In order to enable HMR add `import 'haul/hot';` at the top of the root file - usually `index.ios.js` or `index.android.js`. Then, from within _Developer menu_ tap _Enable Hot Reloading_.
+
+## Manual setup
+Navigate to: [API docs](./API.md)
+
+Jump to: [Examples](#examples)
+
+For advanced project or the ones with different structure, we provide hot API, which allows you to enable HMR regardless of which navigation library you use and how you structure yor code.
```
Comment (satya164) [inline docs/hmr/Setup.md]: `project` -> `projects`
`yor` -> `your`

File: README.md (hunk 4 above)
Comment (zamotany) [inline README.md]: This PR is WIP, not everything is done yet

**Comment by appden** [comment]
@zamotany Thank you continuing to push this forward! I couldn't find where this question might have been answered before, but it appears that this will be using a custom React hot loading solution instead of leveraging [react-hot-loader](https://github.com/gaearon/react-hot-loader). Why is that?

**Comment by zamotany** [comment]
It is using `react-hot-loader`, but it's customized and tweaked to our needs, otherwise it will be of no use.

**Comment by grabbou** [comment]
CC: @Kureev

File: src/HMR/babelPlugin.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
+/**
```
Comment (satya164) [inline src/HMR/babelPlugin.js]: can we change the directory name to `hot`? :)

File: src/server/index.js [hunk 11]
```diff
@@ -27,10 +28,13 @@ const missingBundleMiddleware = require('./middleware/missingBundleMiddleware');
...
 const rawBodyMiddleware = require('./middleware/rawBodyMiddleware');
 
+const WebSocketServer = require('ws').Server;
+
 /**
  * Temporarily loaded from React Native to get debugger running. Soon to be replaced.
  */
-const WebSocketProxy = require('./util/WebsocketProxy.js');
+const webSocketProxy = require('./util/websocketProxy.js');
+const WebSocketDebuggerProxy = require('./util/WebsocketDebuggerProxy.js');
```
Comment (satya164) [inline src/server/index.js]: nit: remove extension

File: src/HMR/client/utils.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+/* eslint-disable import/no-extraneous-dependencies */
+/* eslint-disable no-unused-expressions */
+
+// $FlowFixMe
+import { Platform } from 'react-native';
+
+export default function resetRedBox() {
+  if (Platform.OS === 'ios') {
+    const RCTRedBox = require('NativeModules').RedBox;
+    RCTRedBox && RCTRedBox.dismiss && RCTRedBox.dismiss();
```
Comment (satya164) [inline src/HMR/client/utils.js]: Is it possible that `RCTRedBox` and `RCTExceptionsManager` may not exist?

nit: let's remove the `RCT` prefix. It's a convention on Objective-C side. looks weird in JS

File: src/HMR/client/importScriptsPolyfill.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+ * hot update and will use `importScripts` which are available in WebWorkers, so we need to
+ * provide implementation for it.
+ * 
+ * Native `importScripts` is synchronous, however we can't do that, so this polyfill
+ * is async and returns a Promise.
+ */
+global.importScripts =
+  global.importScripts ||
+  (importPath =>
+    new Promise((resolve: Function, reject: Function) => {
```
Comment (satya164) [inline src/HMR/client/importScriptsPolyfill.js]: What's the reason we use `XMLHttpRequest` instead of `fetch`?

File: src/HMR/client/hotPatch.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    if (typeof uniqueLocalName !== 'string' || typeof fileName !== 'string') {
+      return;
+    }
+
+    const id: string = fileName + '#' + uniqueLocalName; // eslint-disable-line prefer-template
+    if (!idsByType.has(type) && hasCreatedElementsByType.has(type)) {
+      if (!didWarnAboutID[id]) {
+        didWarnAboutID[id] = true;
+        // eslint-disable-next-line
+        const baseName = fileName.replace(/^.*[\\\/]/, '');
```
Comment (satya164) [inline src/HMR/client/hotPatch.js]: Not clear what this does. Maybe add a comment?

Also why disable ESLint?

File: src/HMR/babelPlugin.js (hunk 10 above)
Comment (Kureev) [inline src/HMR/babelPlugin.js]: 🔥

File: src/HMR/client/hotPatch.js (hunk 14 above)
Comment (zamotany) [inline src/HMR/client/hotPatch.js]: No idea, didn't wrote it.

File: src/HMR/client/utils.js (hunk 12 above)
Comment (zamotany) [inline src/HMR/client/utils.js]: Copied from native hot client.

File: src/server/middleware/hotMiddleware.js [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+  // Stats has children, multiple bundles
+  if (stats.children && stats.children.length) return stats.children;
+
+  // Not sure, assume single
+  return [stats];
+}
+
+function buildModuleMap(modules: Object[]): Object {
+  const map = {};
```
Comment (satya164) [inline src/server/middleware/hotMiddleware.js]: Maybe use`reduce`? xD

https://twitter.com/satya164/status/897153823415955456

Comment (zamotany) [inline src/server/middleware/hotMiddleware.js]: Not my code -> low priority. https://github.com/glenjamin/webpack-hot-middleware

File: src/hot/client/importScriptsPolyfill.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  global.importScripts ||
+  (importPath =>
+    fetch(importPath).then(response => response.text()).then(body => {
+      try {
+        // eslint-disable-next-line no-eval
+        eval(body);
+      } catch (error) {
+        return Promise.reject(error);
+      }
+      return Promise.resolve();
```
Comment (satya164) [inline src/hot/client/importScriptsPolyfill.js]: You don't need this

File: src/hot/client/importScriptsPolyfill.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+ * provide implementation for it.
+ * 
+ * Native `importScripts` is synchronous, however we can't do that, so this polyfill
+ * is async and returns a Promise.
+ */
+global.importScripts =
+  global.importScripts ||
+  (importPath =>
+    fetch(importPath).then(response => response.text()).then(body => {
+      try {
```
Comment (satya164) [inline src/hot/client/importScriptsPolyfill.js]: This block can be reduced to

```js
then(body => eval(body))
```

**Comment by satya164** [comment]
Not really, but let's merge this!

We should release a beta with this and see how it works.

**Comment by grabbou** [comment]
I have merged this pull request - nice job! Before releasing, I will try it in a sample project and see how it goes. ETA for this is either later today or tomorrow (don't quote me on that)
